### PR TITLE
Implement proto.Models for OpenAPI v3

### PR DESCRIPTION
**Pull request by alexzielenski**
This PR adds the ability to parse an gnostic `openapi_v3.Document` into types conforming to the interface required by kubectl to power the `explain` feature.

This is a temporary measure as we prepare to replace usage of the gnostic types with the Kube-OpenAPI's `spec3` types in a future release.

### CODE REVIEW

File: pkg/util/proto/openapi_test.go [hunk 1]
```diff
@@ -28,6 +28,7 @@ import (
 
 var fakeSchema = testing.Fake{Path: filepath.Join("testdata", "swagger.json")}
 var fakeSchemaNext = testing.Fake{Path: filepath.Join("testdata", "swagger_next.json")}
+var fakeSchemaV300 = testing.Fake{Path: filepath.Join("testdata", "openapi_v3_0_0")}
```
Comment (apelisse) [inline pkg/util/proto/openapi_test.go:31]: 🤖 😎

File: pkg/util/proto/document_v3.go [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+
+	components := doc.GetComponents()
+	if components == nil {
+		return &definitions, nil
+	}
+
+	schemas := components.GetSchemas()
+	if schemas == nil {
+		return &definitions, nil
+	}
```
Comment (apelisse) [inline pkg/util/proto/document_v3.go:43]: Maybe I'm wrong, but if gnostic still works the same, it usually works well when you call methods on nil pointers. So you can shorten that to:
```suggestion
	schemas := components.GetComponents().GetSchemas()
	if schemas == nil {
		return &definitions, nil
	}
```

File: pkg/util/proto/testing/openapi.go [hunk 3]
```diff
@@ -52,6 +56,35 @@ func (f *Fake) OpenAPISchema() (*openapi_v2.Document, error) {
 	return f.document, f.err
 }
 
+func (f *Fake) OpenAPIV3Discovery() (*handler3.OpenAPIV3Discovery, error) {
```
Comment (apelisse) [inline pkg/util/proto/testing/openapi.go:59]: I noticed that you haven't used this, how do you plan on using this?

File: pkg/util/proto/testing/openapi.go [hunk 4]
```diff
@@ -52,6 +56,35 @@ func (f *Fake) OpenAPISchema() (*openapi_v2.Document, error) {
...
+	_, err := os.Stat(f.Path)
+	if err != nil {
+		return nil, err
+	}
+	spec, err := ioutil.ReadFile(filepath.Join(f.Path, groupVersion+".json"))
+	if err != nil {
+		return nil, err
+	}
+	return openapi_v3.ParseDocument(spec)
+}
```
Comment (apelisse) [inline pkg/util/proto/testing/openapi.go:86]: This one doesn't cache the result, which is sort of the promise of the `Fake` object, maybe we need a `FakeV3`?

File: pkg/util/proto/document_v3.go [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+	"reflect"
+	"strings"
+
+	openapi_v3 "github.com/google/gnostic/openapiv3"
+	"gopkg.in/yaml.v3"
+)
+
+// Temporary parse implementation to be used until gnostic->kube-openapi conversion
+// is possible.
+func NewOpenAPIV3Data(doc *openapi_v3.Document) (Models, error) {
```
Comment (apelisse) [inline pkg/util/proto/document_v3.go:30]: I'm not excited to see more of this, but hopefully we can get rid of it altogether soon.

File: pkg/util/proto/testing/openapi.go (hunk 4 above)
Comment (alexzielenski) [reply pkg/util/proto/testing/openapi.go]: I added caching to this method of the `Fake`.

On a slightly related note,  I did end up adding the following mock class to `client-go` which I found more useful in my testing:

https://github.com/alexzielenski/kubernetes/blob/cache-busting/staging/src/k8s.io/client-go/util/testing/fake_openapi_handler.go#L40

Maybe it is more appropriate to add to kube-openapi, but I did not use it there. I use this mock server so far in tests for `client-go`, `kubectl`, and `kube-aggregator`

File: pkg/util/proto/testing/openapi.go [hunk 6]
```diff
@@ -32,6 +34,10 @@ type Fake struct {
 	once     sync.Once
 	document *openapi_v2.Document
 	err      error
+
+	v3DocumentsLock sync.Mutex
```
Comment (apelisse) [inline pkg/util/proto/testing/openapi.go:38]: You have members of that struct that only make sense for v3 and members that only make sense for v2. Can you split into two different structs?

Comment (alexzielenski) [reply pkg/util/proto/testing/openapi.go:38]: yeah fair enough

**Comment by apelisse** [comment]
Thanks
/lgtm
/approve

File: pkg/util/proto/document_v3.go (hunk 5 above)
Comment (liubog2008) [reply pkg/util/proto/document_v3.go:30]: Does this conversion is ready now ?

**Pull request merged**